lib/compat/wordpress-5.9/get-global-styles-and-settings.php

```diff
@@ -23,7 +23,7 @@
 	 *
 	 * @return array The settings to retrieve.
 	 */
-	function wp_get_global_settings( $path = array(), $context = array() ) {
+	function gutenberg_get_global_settings( $path = array(), $context = array() ) {
```

Comment:
Thanks for getting this up so we can have a looksy @andrewserong !!

Looks like a similar approach to the one we suggested over at https://github.com/WordPress/gutenberg/pull/38671#issuecomment-1034433841 and over at https://github.com/WordPress/gutenberg/pull/38816/files#r808663126, where I'm looking to tackle something similar.

As mentioned over at https://github.com/WordPress/gutenberg/pull/38816/files#r808663126, I'm starting to wonder if we should move changes to this file over to `compat/6.0` in a "compatibility preparation branch or something" since we'll have perform some further trickery on it.

By "trickery" I mean updating methods such as `{wp|gutenberg}_get_global_stylesheet()` to use `WP_Theme_JSON_Resolver_Gutenberg` instead of `WP_Theme_JSON_Resolver`, so it picks up the compat/6.0 changes in `WP_Theme_JSON_Gutenberg` for example.

I'm still lacking a lot of the context and knowledge of the approaches that were taken for 5.9, so maybe @oandregal @jorgefilipecosta @aristath or @youknowriad might have some tips for us.

Reply:
I'm looking at this one and I'm going to push changes in the next few hours.

Reply:
Hey Andrew, I've heavily edited the commits in this PR to show how can we override global styles code. Sorry about the intrusion. :bowing_man: 

Here's what we can do to add new things to the global styles code in the server:

1. Update the allowed lists (`PROPERTIES_METADATA`, `VALID_SETTINGS`, etc) with the new settings in `lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php`. This suffices to make the plugin work in WordPress 5.8. See this commit for example https://github.com/WordPress/gutenberg/pull/38039/commits/e2a18908c1071490917fb437a73bcd01f9f0b814
2. Make sure the plugin calls the plugin classes (`WP_Theme_JSON_Resolver_Gutenberg` and `WP_Theme_JSON_Gutenberg`). Because our main interfaces are the `wp_get_global_*` functions, the changes in step 1 are not enough for WordPress 5.9 because those functions are part of core and call the core classes. In WordPress 5.8 it works fine because it uses the plugin classes. How can we override the functions in 5.9? We can't, so we instead call them `gutenberg_get_global_*` in the plugin and use those in all the instances. This commit renames `wp_get_global_settings` to `gutenberg_get_global_settings` and changes all the instances https://github.com/WordPress/gutenberg/pull/38039/commits/cf100a22e52d52bedcc9ecd021d872e6eb84d51b

That's it. Note that we only need the second step once: when this is done, subsequent PRs won't have to do it, they'll only require the changes in step 1.

Andrew, I've noticed you have added `layout.position` to the properties metadata, so I understood that'd be a valid style going forward. I went ahead I repeated the process I did with settings:

1. Update the allowed lists in `lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php`. https://github.com/WordPress/gutenberg/pull/38039/commits/3e935bd4c9571d2272c66b248de6b2d118f7a7fd
2. Make sure the plugin functions call the plugin classes: https://github.com/WordPress/gutenberg/pull/38039/commits/f20fa0c5262d1e8797da3dc377e7e349fa3e15d8 In testing this I've noticed a bug that I've fixed at https://github.com/WordPress/gutenberg/pull/38039/commits/ca41acfa4bea1655a590159563581c887a5467d6

Reply:
I've figured there are a few in-flight PRs you all are working on and we don't know which will land first. So I've gone ahead and extracted the changes done in step 2 (both settings & styles) to a standalone PR so we can fast-track them and your PRs focus on feature work https://github.com/WordPress/gutenberg/pull/38883

Reply:
cc @andrewserong @ramonjd @aaronrobertshaw @ajlende @scruffian for awareness of this thread, as I know you all have been working on this area.

Reply:
Huge kudos for helping on this @oandregal! 🙇 And for the great explanations. It's a massive help. Thank you!!

Reply:
That's fantastic, thanks so much for jumping in and committing to the PR @oandregal, much appreciated! We'll take a look at #38883 today 😀